### PR TITLE
feat: allow cancellation of workspace symbols requests

### PR DESCRIPTION
**Pull request by tatchi**

### CODE REVIEW

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 1]
```diff
@@ -660,17 +660,38 @@ let workspace_symbol server (state : State.t) (params : WorkspaceSymbolParams.t)
...
+            match cancel with
+            | None ->
+              fun f ->
+                let+ res = f () in
+                (res, Fiber.Cancel.Not_cancelled)
+            | Some token ->
+              let on_cancel () = Lev_fiber.Thread.cancel task in
+              fun f -> Fiber.Cancel.with_handler token ~on_cancel f
+          in
+          maybe_cancel @@ fun () -> Lev_fiber.Thread.await task
```
Comment (tatchi) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:680]: Might be a bit too early to say but I've the impression that this code will be often used with cancellation. Would it make sense to extract it? Something like:

```ocaml
let make_with_cancellation cancel ~on_cancel =
  match cancel with
  | None ->
    fun f ->
      let+ res = f () in
      (res, Fiber.Cancel.Not_cancelled)
  | Some token ->
    fun f -> Fiber.Cancel.with_handler token ~on_cancel f
```

Which could be used like so:

```ocaml
let* res, cancel =
  match task with
  | Ok task ->
    let* cancel = Server.cancel_token () in
    let on_cancel () = Lev_fiber.Thread.cancel task in
    let with_cancellation = make_with_cancellation cancel ~on_cancel in
    with_cancellation @@ fun () -> Lev_fiber.Thread.await task
  | Error `Stopped -> Fiber.never
```

Comment (rgrinberg) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml:680]: If you can clean up the code with this then sure.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 2]
```diff
@@ -660,17 +660,38 @@ let workspace_symbol server (state : State.t) (params : WorkspaceSymbolParams.t)
     let+ symbols_results =
       let task =
         Lev_fiber.Thread.task thread ~f:(fun () ->
+            let+ () = Fiber.return () in
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:663]: why do we need this line?

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 3]
```diff
@@ -660,17 +660,38 @@ let workspace_symbol server (state : State.t) (params : WorkspaceSymbolParams.t)
...
+        | Ok task ->
+          let* cancel = Server.cancel_token () in
+          let maybe_cancel =
+            match cancel with
+            | None ->
+              fun f ->
+                let+ res = f () in
+                (res, Fiber.Cancel.Not_cancelled)
+            | Some token ->
+              let on_cancel () = Lev_fiber.Thread.cancel task in
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:677]: That looks good, but I think we can do even better. Why don't we check if the cancellation token has been fired before reading every cmt file? If it has been fired, we can terminate the work early.

File: ocaml-lsp-server/src/ocaml_lsp_server.ml (hunk 1 above)
Comment (tatchi) [reply ocaml-lsp-server/src/ocaml_lsp_server.ml:679]: I'll eventually do that in another PR :)

**Review by rgrinberg — APPROVED** [review]

**Comment by rgrinberg** [comment]
Merged.

**Pull request closed without merging**